core/src/main/java/org/apache/spark/shuffle/sort/ShuffleInMemorySorter.java

```diff
@@ -69,7 +72,11 @@ public int numRecords() {
     return pos;
   }
 
-  public void reset() {
+  public void freeMemoryAndResetPointerArray() {
+    if (consumer != null) {
+      consumer.freeArray(array);
+      this.array = consumer.allocateArray(initialSize);
```

Comment:
Considering that another concurrent task could grab all the memory freed by `consumer.freeArray(array)`.

Another option could be manually release the array, then call releaseMemory() by passing the different of the two.